### ISSUE TITLE
tables are not formatted nicely

### ISSUE THREAD

**Issue by funnes** [rvben/rumdl#129]
`rumdl --no-config fmt` (version 0.0.169) turns this:

``` md
| Header 1 | Header 2 |
| -------- | -------- |
| Cell 1   | Cell 2
```

into this:

``` md
| Header 1 | Header 2 |
| --------|-------- |
| Cell 1 | Cell 2 |
```

Is this expected behavior?

I'd expect it to be formatted like this:

``` md
| Header 1 | Header 2 |
| -------- | -------- |
| Cell 1   | Cell 2   |
```

**Comment by rvben** [comment]
Hi @funnes,

That looks like a bug! It should not be reformatted, just fixed. I'll look into it.

**Comment by funnes** [comment]
Hi 🙂 , thanks for the quick reply.

The most tedious part for me, when writing markdown files, is formatting tables. When trying out rumdl I was hoping it would take care of table formatting completely. I was hoping it would turn this:

``` md
|lorem ipsum|dolor|
|-|-|
|sit|amet consetetur sadipscing elitr|
```

into this:

``` md
| lorem ipsum | dolor                            |
| ----------- | -------------------------------- |
| sit         | amet consetetur sadipscing elitr |
```

Is this something you are interested in supporting?

**Comment by rvben** [comment]
Hi @funnes,

I spent some effort in the past in that direction, but got demotivated when finding out that there are a lot of characters that have widths that do not just fill 1 space, but might be 2 even. To properly format that, it would be sometimes needed to know which font is used or how an emoticon is shown in an editor. 

While thinking about this yesterday, I realized that it might help in more than 90% of the usecases to provide auto-formatting for tables without special characters. I need to think about how to provide the best user experience, but it's definitely something I'd be willing to consider.

Kind regards,

Ruben

**Comment by funnes** [comment]
For me, it would be fine if it assumed that every character has the same width.

**Comment by rvben** [comment]
Hi @funnes,

I've added the MD060 rule specifically for this purpose, but as it's new and experimental it's going to be opt-in for now. (First released yesterday in v0.0.171)
Please enable it and have a look if it does what you're looking for.

```toml
[MD060]
enabled = true
```

Looking forward to your feedback.

**Comment by funnes** [comment]
Nice, with this config I have what I was looking for:

``` toml
[MD060]
enabled = true
style = "aligned"
```

slight nit: tables with empty header are not formatted:

```markdown
|||
|-|-|
|lorem|ipsum|
```

GFM seems to allow tables with an empty header:

|||
|-|-|
|lorem|ipsum|

**Comment by funnes** [comment]
not sure if this is intended behavior but this:

```markdown
|lorem|ipsum|
|--------------|-|
|dolor|sit|
```
turns into this:
```markdown
| lorem          | ipsum |
| -------------- | ----- |
| dolor          | sit   |
```

**Comment by rvben** [comment]
Thanks for the feedback. I'll have a look

**Comment by rvben** [comment]
The fixes should now be released as part of 0.0.172. Let me know if you hit anything else.

**Comment by funnes** [comment]
Amazing, thank you!

**Comment by tuckerkevin** [comment]
I just found out about rumdl and have been playing around with it in Zed and the CLI. There are some additional odd behaviors with the formatting in Zed but I'll restrict this feedback to the CLI to keep things straight...

Try `rumdl check` and `rumdl fmt` on this file:

```md
# table examples

## unformatted

I am not seeing any `MD060` here!?

|default|left|center|right|
|---|:--|:-:|--:|
|default|left|center|right|
|left align|align|align|align|
|me|me|me|me|
|longer text|longer text|longer text|longer text|

|||||
|---|:--|:-:|--:|
|default|left|center|right|
|left align|align|align|align|
|me|me|me|me|

## expected formatting, maybe?

This shows me `MD060`. Other formatters seem to handle center and right aligned columns.

| default     | left        |   center    |       right |
| ----------- | :---------- | :---------: | ----------: |
| default     | left        |   center    |       right |
| left align  | align       |    align    |       align |
| me          | me          |     me      |          me |
| longer text | longer text | longer text | longer text |

|            |       |        |       |
| ---------- | :---- | :----: | ----: |
| default    | left  | center | right |
| left align | align | align  | align |
| me         | me    |   me   |    me |


## commented out unformatted

This shows me `MD058` and `MD060`, arguably shouldn't.

<!--
| default     | left        |   center    |       right |
| ----------- | :---------- | :---------: | ----------: |
| default     | left        |   center    |       right |
| left align  | align       |    align    |       align |
| me          | me          |     me      |          me |
| longer text | longer text | longer text | longer text |

|            |       |        |       |
| ---------- | :---- | :----: | ----: |
| default    | left  | center | right |
| left align | align | align  | align |
| me         | me    |   me   |    me |
-->

## final newline (unrelated issue)

`rumdl fmt` strips the final newline on this line, then shows `MD047` on subsequent check.
```

(edit: expanded out example to include tables with and without headings)

**Comment by rvben** [comment]
Thanks @tuckerkevin for your example. I'll have a look. MD060 is definitely still a bit in an alpha/beta state, I hope I have the time soon to iron out the bugs.

**Comment by rvben** [comment]
Fixes released as part of 0.0.175. Thanks again, @tuckerkevin and @funnes.

**Comment by tuckerkevin** [comment]
Seems to be working much better, but still not warning or fixing a completely unformatted table unless a space is added somewhere (see my first section examples above). At least in my testing.

(But thank you, working much more as expected now.)

**Comment by tuckerkevin** [comment]
I found another weird one...

```md
## longer headings

| a long heading | another long heading |
| -------------- | -------------------- |
| a              | 1                    |
| b b            | 2                    |
| c c c          | 3                    |
```

It doesn't like that for some reason and changes it to:

```md
| a long heading | another long heading |
| -------------- | -------------------- |
| a | 1 |
| b b | 2 |
| c c c | 3 |
```

**Comment by rvben** [comment]
Hi @tuckerkevin,

I can't reproduce the issue in 0.0.176. Can you confirm it's fixed?

**Comment by tuckerkevin** [comment]
For me it is still wanting to change the left to right...
```
58 | a long heading | another long heading |                         58 | a long heading | another long heading |
59 | -------------- | -------------------- |                         59 | -------------- | -------------------- |
60 | a              | 1                    |                         60 | a | 1 |
61 | b b            | 2                    |                         61 | b b | 2 |
62 | c c c          | 3                    |                         62 | c c c | 3 |
```

~~(and still not wanting to touch these)~~
(Edit: understood now, this is likely being detected as "tight" style and left alone.)

```md
|default|left|center|right|
|---|:--|:-:|--:|
|default|left|center|right|
|left align|align|align|align|
|me|me|me|me|
|longer text|longer text|longer text|longer text|

|||||
|---|:--|:-:|--:|
|default|left|center|right|
|left align|align|align|align|
|me|me|me|me|
```

**Comment by zx8** [comment]
@tuckerkevin This seemed to fix it for me:

```
[MD013]
line-length = 9999999999
```

**Comment by tuckerkevin** [comment]
(Update: understood now, the space is causing this to be detected as "compact" style instead of tight, "aligned" was not explicitly called for in settings.)

If I put a single space in, it acts different:

```
 7 | default|left|center|right|                                       7 | default     | left        |   center    |       right |
 8 |---|:--|:-:|--:|                                                  8 | ----------- | :---------- | :---------: | ----------: |
 9 |default|left|center|right|                                        9 | default     | left        |   center    |       right |
10 |left align|align|align|align|                                    10 | left align  | align       |    align    |       align |
11 |me|me|me|me|                                                     11 | me          | me          |     me      |          me |
12 |longer text|longer text|longer text|longer text|                 12 | longer text | longer text | longer text | longer text |
```

(Which looks correct to me.)

**Comment by tuckerkevin** [comment]
> [@tuckerkevin](https://github.com/tuckerkevin) This seemed to fix it for me:
> 
> ```
> [MD013]
> line-length = 9999999999
> ```

Interesting... I have MD013 disabled. But I'll check.

(Update: no change in any of my results with that, whether MD013 enabled or not.)

**Comment by zx8** [comment]
Ah right, was worth a shot!

**Comment by rvben** [comment]
I think I found it. In some cases the style detection would conclude the style of the table is compact and format the rest of the table as such.

**Comment by tuckerkevin** [comment]
@rvben Ok, so I see you have documentation for MD060 up now. I didn't realize you had compact and tight styles too. So that explains my confusion about my "tight" example not being reformatted if a style was not specified. But yeah, I guess "any" is miss-detecting aligned in some cases and reformatting as compact. That said, I think I can just set `style = "aligned"` and get my expected results. Sorry for the confusion.

**Comment by rvben** [comment]
The table style detection needed some improvements too, so thanks for helping me realize that. 

Let me know if you hit anything else.
